### ISSUE TITLE
@jit ignores signature for return type

### ISSUE THREAD

**Issue by shoyer** [numba/numba#1087]
```python
import numpy as np
from numba import jit

@jit("float32(float32)", nopython=True)
def f(x):
    return x
```

The return value was definitely truncated to 32 bits internally at some point, but the returned value is always a Python float, not `np.float32`:
```
In [22]: f(3.2)
Out[22]: 3.200000047683716

In [23]: type(f(3.2))
Out[23]: float

In [24]: type(f(np.float32(3.2)))
Out[24]: float
```

**Comment by pitrou** [comment]
Indeed, the return type doesn't make the returned object a Numpy scalar. I frankly don't know if it's a good idea to change this behaviour or not; I wonder how many people use Numpy scalars in the real world (as opposed to the corresponding built-in types).

**Comment by seibert** [comment]
Completely coincidentally, I discovered this behavior last week and it actually turned out to be useful for me.  (I'm using Numba to convert rows of an ndarray with a structured dtype into tuples of Python scalars.)

There is a performance implication to returning numpy scalars:
``` python
n [9]: import numpy as np

In [10]: a, b = 1.0, 2.0

In [11]: x, y = np.float32(1), np.float32(2)

In [12]: %timeit a + b
10000000 loops, best of 3: 64.6 ns per loop

In [13]: %timeit x + y
10000000 loops, best of 3: 98.1 ns per loop
```

That said, I can see how it is confusing to the user if they explicitly ask for `np.float32` and get a Python float.

The root problem is that we have no way to handle (or give the user control over) multiple mappings from a Numba type to Python type when boxing the return value.  We assume there is a one-to-one mapping, when the situation is potentially ambiguous in this case for numeric scalar types.

Maybe we need a Numba "float" type (that gets boxed as a Python float) that is distinct from "float32" and "float64" types (that get boxed as Numpy float types)?

**Comment by shoyer** [comment]
My use-case is trying to write fast nan-skipping aggregation functions that are consistent with, but faster than, e.g., `np.nanmean`. It's useful to be able to preserve boxed types here:
```
In [3]: x = np.arange(5, dtype='float32')

In [4]: np.nanmean(x)
Out[4]: 2.0

In [5]: np.nanmean(x).dtype
Out[5]: dtype('float32')

In [6]: np.mean(x).dtype
Out[6]: dtype('float32')
```

**Comment by shoyer** [comment]
@pitrou As for numpy scalars vs built-in types, numpy's almost always return numpy scalars instead of built-in types (e.g. from `np.mean`) so users already know to expect it. NumPy actually makes its scalars pretty transparent, because they have a matching `__repr__` with the built-in types.

**Comment by max-sixty** [comment]
FYI numbagg has seen a resurgence in development, and this would still be something we'd been keen to use if it were available.

(I'm not as involved in the ecosystem as others on this thread are, but also matching numpy's return types makes a lot of sense to me)
